### ISSUE TITLE
Room schema json is not updated after updating tables

### ISSUE THREAD

**Issue by rHlushan** [gradle/android-cache-fix-gradle-plugin#157]
If we will update DB (f.e. column's names) but keep schema version number same - schema json file won't be updated.
But originally it will, even without clean build.

**Comment by runningcode** [comment]
@rHlushan what version of AGP, Kotlin Gradle Plugin and Gradle are you using?
Can you also share how you are specifying the room schema?

**Comment by rHlushan** [comment]
AGP version = 4.1.2
Korlin version = 1.4.31
Gradle: distributionUrl = https\://services.gradle.org/distributions/gradle-6.8-all.zip
Android cache fix Gradle plugin version = 2.2.0


specifying the room schema:

added:
room {
    schemaLocationDir = file("$projectDir/room_db_schemas")
}


removed:
android {
    kapt {
        arguments {
            arg("room.schemaLocation", "$projectDir/room_db_schemas".toString())
        }
    }
}

**Comment by runningcode** [comment]
Thanks. @rHlushan does it work as expected if you specify the room schema within the `kapt` extension?

**Comment by rHlushan** [comment]
Sorry, u mean:
add Android cache fix Gradle plugin
and

remove
room {
schemaLocationDir = file("$projectDir/room_db_schemas")
}

but add
android {
kapt {
arguments {
arg("room.schemaLocation", "$projectDir/room_db_schemas".toString())
}
}
}

?

**Comment by runningcode** [comment]
Yeah, could you try that?

**Comment by runningcode** [comment]
@ghale do you have any thoughts here?

**Comment by rHlushan** [comment]
Tried: yes it works as expected.
But is this way correct?

**Comment by rHlushan** [comment]
Oh.
Not so easy.

removed:
room {
schemaLocationDir = file("$projectDir/room_db_schemas")
}

but added:
android {
kapt {
arguments {
arg("room.schemaLocation", "$projectDir/room_db_schemas".toString())
}
}
}

Precondition: json version №1.

Screnario:
0) builded app from clean

1) changed some column name
2) builded it again
    **json was updated** / actual json version: №2 / correct
    
3) returned column name as it was before
4) builded it again
    **json was not updated again to previous version №1** / actual json version: №2 / **not correct**
    
5) changed some column name(not as in step 1)
6) builded it again
    **json was updated** / actual json version: №3 / correct    
    
7) returned column name to original name()
8) builded it again
    **json was not updated again to previous version №1** / actual json version: №3 / **not correct**

**Comment by ghale** [comment]
It's not 100% clear to me what is incorrect.

In 3 above, my understanding is that you updated the column annotation to change the column name back and then also updated the database annotation to set the version back to 1.  Is this correct?  

If so, what are you referring to as the "actual json version" in 4 above?  Is that the name of the schema file or the version in the schema file?  For instance, are you implying that 1.json now has "2" as the value for database.version?  Or are you implying that because 2.json still exists in the schema output directory, this is incorrect?

**Comment by rHlushan** [comment]
Sorry if it was nit clear.
Let me try in other way:

Version number was not changed at all.
By json version I mean json content(like columns description, database hash, indices description, ....), but not json mane itself.

Example:

Precondition:

database @Database(
        .....
        version = 1,
        exportSchema = true
)

some entity: @ColumnInfo(name = "name")

filename: 1.json / version of content №1("columnName": "name")

Screnario:
0) builded app from clean

1) changed column name @ColumnInfo(name = "name_1")

2) builded it again
json content was updated / actual json content version: №2("columnName": "name_1") / correct

3) returned column name to original name @ColumnInfo(name = "name")

4) builded it again
json content was not updated again to previous content version №1("columnName": "name") / actual json content version: №2(still "columnName":  "name_1") / not correct

5) changed some column name(not as in step 1) @ColumnInfo(name = "name_2")

6) builded it again
json content was updated / actual json content version: №3("columnName": "name_2") / correct

7) returned column name to original name @ColumnInfo(name = "name")

8) builded it again
json content was not updated again to content version №1("columnName": "name") / actual json version: №3("columnName": "name_2") / not correct

**Comment by ghale** [comment]
Ah, I see.  I think I know what's going on.  Can you confirm that the file in `<project_dir>/build/roomSchemas/kapt<variant_name>Kotlin` has the correct contents?

**Comment by rHlushan** [comment]
after step 8 above:

this folder is empty

build again(not clean) -> folder is empty

changed column name again -> build again(not clean) -> folder is still empty, but json content was successfully changed

**Comment by ghale** [comment]
I suspect the changes in #181 to fix the problem described in #183 may help with this issue.  If you could try plugin version 2.4.1 and confirm whether you're still seeing issues, that would be great.

**Comment by mahmoud-elmorabea** [comment]
Hey @ghale,

Thanks for the fix! 

Unfortunately we are still running into a problem.

I've tested with 2.4.1 and it seems like the changes in #181 does fix the problem if a Room `@Entity` changes or a new one is added, but in the case of an `@Entity` being removed, I can still see the wrong merge happening and the generated JSON file contains the wrong version.

The same issue also happens with our current version of cache-fix 2.3.3

We are also running Gradle 7.0 and AGP 4.2.2

**Comment by mahmoud-elmorabea** [comment]
Hey @ghale ,

Please ignore my last comment, it turned out that is not the problem that is happening.

After lots of trials I came to the conclusion that one of the kotlin compile tasks had its output cached and for some reason that output was causing the `mergeRoomSchemaLocations` task to generate an incorrect schema JSON file (with old version).

Running everything the same with `--no-build-cache` always produced the correct result, however, with cache enabled (both remote or local) the problem persisted.

Changing a random Kotlin file caused the kotlin compile task to rerun and everything was fine again. I honestly have no idea what was wrong with that task output.

I will let you know if I see that again. Sorry about the earlier message though.

**Comment by etiennestuder** [comment]
Note that there is also the option `--rerun-tasks` to force all tasks to be re-executed and put into the cache again.

**Comment by mahmoud-elmorabea** [comment]
Yea, that's true! Thanks for the pointer.

I didn't want to override the cache for that task before I was sure it was the problem though, that's why I was using `--no-build-cache` 😄

**Comment by ghale** [comment]
I'm going to close this for now.  If we find that there are still any remaining problems after the latest changes, we can always re-open or create a new issue.
